shared_model/backend/protobuf/common_objects/amount.hpp

```diff
@@ -27,15 +27,29 @@
 namespace shared_model {
   namespace proto {
     class Amount final : public interface::Amount {
-     private:
-      using RefAmount = detail::ReferenceHolder<iroha::protocol::Amount>;
-
      public:
-      explicit Amount(const iroha::protocol::Amount &amount)
-          : Amount(RefAmount(amount)) {}
+      template <typename AmountType>
+      explicit Amount(AmountType &&amount)
+          : proto_amount_(std::forward<AmountType>(amount)),
+            multiprecision_repr_([this] {
+              const auto offset = 64u;
+              auto times = 3u;
+              boost::multiprecision::uint256_t result;
+              result |= proto_amount_->value().first() << offset * times--;
```

Comment:
Not an issue, but I am not so happy about using post-increment in complex expressions.

Reply:
Explain why

Reply:
Actually current implementation also helps to see that values are shifted only three times.